qiskit/transpiler/passes/scheduling/time_unit_conversion.py

```diff
@@ -71,52 +73,90 @@ def run(self, dag: DAGCircuit):
         if self._durations_provided:
             inst_durations.update(self.inst_durations, getattr(self.inst_durations, "dt", None))
 
+        # The float-value converted units for delay expressions, either all in 'dt'
+        # or all in seconds.
+        expression_durations = {}
+
         # Choose unit
-        if inst_durations.dt is not None:
-            time_unit = "dt"
-        else:
-            # Check what units are used in delays and other instructions: dt or SI or mixed
-            units_delay = self._units_used_in_delays(dag)
-            if self._unified(units_delay) == "mixed":
+        has_dt = False
+        has_si = False
+
+        # We _always_ need to traverse duration expressions to convert them to
+        # a float. But we also use the opportunity to note if they intermix cycles
+        # and wall-time, in case we don't have a `dt` to use to unify all instruction
+        # durations.
+        for node in dag.op_nodes(op=Delay):
+            if isinstance(node.op.duration, expr.Expr):
+                if any(
+                    isinstance(x, expr.Stretch) for x in expr.iter_identifiers(node.op.duration)
+                ):
+                    # If any of the delays use a stretch expression, we can't run scheduling
+                    # passes anyway, so we bail out. In theory, we _could_ still traverse
+                    # through the stretch expression and replace any Duration value nodes it may
+                    # contain with ones of the same units, but it'd be complex and probably unuseful.
+                    self.property_set["time_unit"] = "stretch"
+                    return dag
+
+                visitor = _EvalDurationImpl(inst_durations.dt)
+                duration = node.op.duration.accept(visitor)
+                if visitor.in_cycles():
+                    has_dt = True
+                    # We need to round in case the expression evaluated to a non-integral 'dt'.
+                    rounded_duration = round(duration)
+                    rounding_error = abs(duration - rounded_duration)
+                    if rounding_error > 1e-15:
+                        warnings.warn(
+                            f"Duration is rounded to {rounded_duration:d} [dt] from {duration:f}",
+                            UserWarning,
+                        )
+                    duration = rounded_duration
+                else:
+                    has_si = True
+                if duration < 0:
+                    raise TranspilerError(
+                        f"Expression '{node.op.duration}' resolves to a negative duration!"
+                    )
```

Comment:
I don't think we do this for legacy durations, even though they're representable.  Do we do the check somewhere else (and so should let that somewhere-else handle it), or is it a failure path we should handle for other things?

Fwiw, `TimeUnitConversion` _itself_ probably doesn't care if a duration is negative (the conversions are well defined anyway), but other places would.

Reply:
Oh, if literals would blow up because `Delay` wouldn't accept it, it's fine.

Reply:
Yup, yeah they would.